### ISSUE TITLE
iptables state + mod_aggregate: save -- non-functional

### ISSUE THREAD

**Issue by gladiatr72** [saltstack/salt#20818]
https://github.com/saltstack/salt/blob/develop/salt/states/iptables.py#L465
https://github.com/saltstack/salt/blob/develop/salt/states/iptables.py#L342

I scavenged the iptables modules/state from trunk to try with a 2014.7.1 core--was very interested in the mod_aggregate work that Mr. Greenaway committed back in November.  It works splendidly except for one thing: The 'save' option is currently not propagated.

This may be news to Absolutely No One, but I wanted to be sure to get an issue opened.  Aggregating inserts and appends solves a lot of fanduckery that must ensue to properly order rules.  I think a driving force was to avoid having a file written to disk for every rule in the set (I'm sure this gets ridiculous at a certain point)

There are two things I'd like to see:

. Appends and Inserts get aggregated to the same set
. The rule array (in its entirety) have the save option applied--if it appears once, it applies to all aggregated rules in the state run.  Then we get the do-many-things but only hit the disk once.

Another concept that occurred to me while working on iptables states today:

How much conceptual pain would be involved to create a (meta?)state module that would allow one to be able to grab ahold of an aggregated rule set?  

Right now we can weave the correct requisites together to effectively create a "top" or "bottom" state that can be watched or required, but that's a lot of work (with a firewall ruleset) and bit on the brittle side.  It would be a welcome addition to give such a set Thingness.

**Comment by jfindlay** [comment]
@gladiatr72, thanks for looking at this.  From what I've heard from @basepi, iptables is difficult to state because rule changes can be {tenuously|not} idempotent.  I personally don't have much experience with using that state or with the known or historic issues thereof but am happy to help review any code changes you may have.  Your idea about an abstract state sounds good to me, but, you know, I'm not an expert on stating iptables.

**Comment by garethgreenaway** [comment]
The save doesn't work at all?  The intended functionality was to aggregate all the rules into memory then save once they were all in place, instead of running save after each insert or append.

**Comment by gladiatr72** [comment]
It did not.  I thought that would have been part of the intent but by the time it got to the low-state generation there was no sign of saving.

**Comment by garethgreenaway** [comment]
Odd.  Will take a look.

**Comment by gladiatr72** [comment]
Excellent.  Thanks for your work on this module!

**Comment by garethgreenaway** [comment]
I think I see the issue here.  Will submit a pull request in a few minutes.

**Comment by syphernl** [comment]
Hmm, seems like this hasn't solved it though. Each rule (still) results in multiple calls to `iptables-save`, even those that are already there and don't need saving at all:

```
[INFO    ] Running state [iptables_smtp_allow_127.0.0.1] at time 07:17:04.473438
[INFO    ] Executing state iptables.append for iptables_smtp_allow_127.0.0.1
[INFO    ] Executing command '/sbin/iptables -t filter -N 0x8002710f1e7L' in directory '/home/vagrant'
[INFO    ] Executing command '/sbin/iptables -t filter -A 0x8002710f1e7L -p tcp --dport smtp  --source 127.0.0.1 --jump ACCEPT' in directory '/home/vagrant'
[INFO    ] Executing command '/sbin/iptables-save' in directory '/home/vagrant'
[INFO    ] Executing command '/sbin/iptables -t filter -F 0x8002710f1e7L' in directory '/home/vagrant'
[INFO    ] Executing command '/sbin/iptables -t filter -X 0x8002710f1e7L' in directory '/home/vagrant'
[INFO    ] Executing command '/sbin/iptables-save' in directory '/home/vagrant'
[INFO    ] Executing command '/sbin/iptables-save' in directory '/home/vagrant'
[INFO    ] iptables rule for iptables_smtp_allow_127.0.0.1 already set (/sbin/iptables --wait -t filter -A INPUT  -p tcp --dport smtp  --source 127.0.0.1 --jump ACCEPT) for ipv4
[INFO    ] Completed state [iptables_smtp_allow_127.0.0.1] at time 07:17:04.595840
```

I'm using it on Salt `2014.7.1` with the latest (develop) iptables state & module (via the fileserver) and the [iptables-formula](https://github.com/saltstack-formulas/iptables-formula/).

**Comment by syphernl** [comment]
A run using the "stock" version of `2014.7.1` (without aggregate) results in far less save's:
```
[INFO    ] Running state [iptables_smtp_allow_127.0.0.1] at time 07:25:12.641522
[INFO    ] Executing state iptables.append for iptables_smtp_allow_127.0.0.1
[INFO    ] Executing command '/sbin/iptables -t filter -C INPUT -p tcp --dport smtp --source 127.0.0.1 --jump ACCEPT ' in directory '/home/vagrant'
[ERROR   ] Command '/sbin/iptables -t filter -C INPUT -p tcp --dport smtp --source 127.0.0.1 --jump ACCEPT ' failed with return code: 1
[ERROR   ] output: iptables: Bad rule (does a matching rule exist in that chain?).
[INFO    ] Executing command '/sbin/iptables -t filter -A INPUT -p tcp --dport smtp --source 127.0.0.1 --jump ACCEPT ' in directory '/home/vagrant'
[INFO    ] Executing command '/sbin/iptables-save' in directory '/home/vagrant'
[INFO    ] {'locale': 'iptables_smtp_allow_127.0.0.1'}
[INFO    ] Completed state [iptables_smtp_allow_127.0.0.1] at time 07:25:12.670294
```

**Comment by gladiatr72** [comment]
iptables.mod_aggr() didn't make it into the 2014.7.1 branch

**Comment by syphernl** [comment]
I know, but this was more to denote that aggregate is doing more saves than non-aggregate even though it should be the other way around.

**Comment by garethgreenaway** [comment]
Odd.  Will take a look.

**Comment by gladiatr72** [comment]
Gotcha.

**Comment by garethgreenaway** [comment]
I think I see the issue.  Seems related to the function that checks if the rule already exists.  Looping in @jfindlay since he wrote the code.  Looking at line 466 of the iptables module, that looks backwards to me.  If the --check option isn't available then the code that follows should run right?  Unless I'm reading it wrong.

**Comment by jfindlay** [comment]
@garethgreenaway, the updates I made to the iptables module went into 2015.2, whereas I believe @syphernl is encountering these problems on 2014.7.  On the head of 2015.2, the `if _has_option('--check', family):` is line 509 and it seems to prevent the following code from being run if the `--check` option is unavailable.

**Comment by garethgreenaway** [comment]
@jfindlay He's using the develop version of the iptables module & iptables state with 2014.7.  The updates you made for 2015.2 also got merged into develop, which is where I was seeing them.  The _has_option is going to return True is the --check is available so that if with the call to _has_option will be True when --check is available, so the code below if will run.  If --check isn't available then the code that uses --check, in this case -C will run.

**Comment by jfindlay** [comment]
@garethgreenaway, I see; I though @syphernl was using 2014.7.  I've fixed the code so that it runs `--check` only after checking for it.  Does referencing commit correct the issue?

**Comment by garethgreenaway** [comment]
@jfindlay Looks right to me.  Once it's merged there should be less calls to save when using the aggregate function but will defer to @syphernl

**Comment by syphernl** [comment]
Cool, I'll give it a go tomorrow morning!

**Comment by jfindlay** [comment]
@garethgreenaway, @syphernl, sorry that I broke it :-( so thanks for the feedback.

**Comment by syphernl** [comment]
```
[INFO    ] Running state [iptables_smtp_allow_127.0.0.1] at time 10:41:02.964355
[INFO    ] Executing state iptables.append for iptables_smtp_allow_127.0.0.1
[INFO    ] Executing command '/sbin/iptables -t filter -C INPUT -p tcp --dport smtp --source 127.0.0.1 --jump ACCEPT ' in directory '/home/vagrant'
[ERROR   ] Command '/sbin/iptables -t filter -C INPUT -p tcp --dport smtp --source 127.0.0.1 --jump ACCEPT ' failed with return code: 1
[ERROR   ] output: iptables: Bad rule (does a matching rule exist in that chain?).
[INFO    ] Executing command '/sbin/iptables --wait -t filter -A INPUT -p tcp --dport smtp --source 127.0.0.1 --jump ACCEPT ' in directory '/home/vagrant'
[INFO    ] Executing command '/sbin/iptables-save' in directory '/home/vagrant'
[INFO    ] {'locale': 'iptables_smtp_allow_127.0.0.1'}
[INFO    ] Completed state [iptables_smtp_allow_127.0.0.1] at time 10:41:03.002400
```

Less saves are being done, but it is still saving after each rule rather than at the end when all rules have been set.

Tested with this set in the master config:
```
state_aggregate:
  - iptables
```

with the iptables module + state from 85b4fee in the fileserver and synced with the minion.

**Comment by syphernl** [comment]
Err, I should've tested with the module from602b3983969df67e78d919ccffe8a963e2e408be :-)

That one appears to be the same as with my previous comment:
```
[INFO    ] Running state [iptables_smtp_allow_127.0.0.1] at time 10:47:05.030402
[INFO    ] Executing state iptables.append for iptables_smtp_allow_127.0.0.1
[INFO    ] Executing command '/sbin/iptables -t filter -C INPUT -p tcp --dport smtp --source 127.0.0.1 --jump ACCEPT ' in directory '/home/vagrant'
[ERROR   ] Command '/sbin/iptables -t filter -C INPUT -p tcp --dport smtp --source 127.0.0.1 --jump ACCEPT ' failed with return code: 1
[ERROR   ] output: iptables: Bad rule (does a matching rule exist in that chain?).
[INFO    ] Executing command '/sbin/iptables --wait -t filter -A INPUT -p tcp --dport smtp --source 127.0.0.1 --jump ACCEPT ' in directory '/home/vagrant'
[INFO    ] Executing command '/sbin/iptables-save' in directory '/home/vagrant'
[INFO    ] {'locale': 'iptables_smtp_allow_127.0.0.1'}
[INFO    ] Completed state [iptables_smtp_allow_127.0.0.1] at time 10:47:05.071478
```

**Comment by syphernl** [comment]
Second run (if the rule exists):
```
[INFO    ] Running state [iptables_smtp_allow_127.0.0.1] at time 10:48:54.641675
[INFO    ] Executing state iptables.append for iptables_smtp_allow_127.0.0.1
[INFO    ] Executing command '/sbin/iptables -t filter -C INPUT -p tcp --dport smtp --source 127.0.0.1 --jump ACCEPT ' in directory '/home/vagrant'
[INFO    ] Executing command '/sbin/iptables-save' in directory '/home/vagrant'
[INFO    ] Executing command '/sbin/iptables-save' in directory '/home/vagrant'
[INFO    ] iptables rule for iptables_smtp_allow_127.0.0.1 already set (/sbin/iptables --wait -t filter -A INPUT  -p tcp --dport smtp --source 127.0.0.1 --jump ACCEPT) for ipv4
[INFO    ] Completed state [iptables_smtp_allow_127.0.0.1] at time 10:48:54.749536
```

**Comment by garethgreenaway** [comment]
Okay.  Will take a look at why it's saving multiple times when the rule exists.  Thanks for the logs!

**Comment by garethgreenaway** [comment]
Looking at the code a lot of the calls to iptables-save aren't actually writing to disk, but are used to see the rules that are currently in memory and compare them to those saved on disk, to determine whether it needs to save them to disk.

**Comment by gladiatr72** [comment]
Is there potential for this to make it into the upcoming 2015 release?

On Mon, Mar 9, 2015 at 5:55 PM, jfindlay <notifications@github.com> wrote:

> Closed #20818 <https://github.com/saltstack/salt/issues/20818> via 82e020e
> <https://github.com/saltstack/salt/commit/82e020ed3adcf0be840d8bba4af7f059a49e438f>
> .
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/saltstack/salt/issues/20818#event-248945131>.
>



-- 
You know, I used to think it was awful that life was so unfair. Then I
thought, wouldn't it be much worse if life were fair, and all the terrible
things that happen to us come because we actually deserve them? So, now I
take great comfort in the general hostility and unfairness of the universe.

**Comment by rallytime** [comment]
@gladiatr72 Yep! This fix (#21321) was merged into the 2015.2 branch, so it will be in the upcoming 2015.2.0 release.
